src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/DiagramViewModel.java

```diff
@@ -290,15 +291,50 @@ public void setHiddenNodes(Set<Integer> nodes) {
         hiddenNodesChangedEvent.fire();
     }
 
-    public FilterChain getSequenceFilterChain() {
-        return filterChain;
+    private void setFilterChain(FilterChain newFC, FilterChain customFC) {
+        if (filterChain != null) {
+            filterChain.getChangedEvent().removeListener(filterChainChangedListener);
+        }
+        if (newFC.getName().equals(customFC.getName())) {
+            filterChain = customFC;
+        } else {
+            filterChain = newFC;
+        }
+        filterChain.getChangedEvent().addListener(filterChainChangedListener);
+    }
+
+    public void initFiltersFromModel(DiagramViewModel other) {
+        customFilterChain.clearFilters();
+        customFilterChain.addFilters(other.getCustomFilterChain().getFilters());
+        setFilterChain(other.getFilterChain(), customFilterChain);
+    }
+
+    void activateModel() {
+        FilterChainProvider provider = Lookup.getDefault().lookup(FilterChainProvider.class);
+        if (provider != null) {
+            // link the Filters window with this model
+            provider.setFilterChainSelectionChangedListener(l -> {
+                // this function is called when user selects a different filter profile for this model
+                setFilterChain(provider.getFilterChain(), customFilterChain);
+            });
+            provider.setCustomFilterChain(customFilterChain);
+            provider.selectFilterChain(filterChain);
+        }
     }
 
+    // called when the filter in filterChain changed, but not filterChain itself
     private void filterChanged() {
```

Comment:
After applying this PR, `DiagramViewModel::filterChanged()` is fired every time a new graph in a group is viewed. This is not a functional bug, but it causes the expensive `DiagramViewModel::rebuildDiagram()` to be called twice in that scenario (the other call comes from `DiagramViewModel::changed()`). Would it be possible to arrange the code so that `DiagramViewModel::rebuildDiagram()` is only called once when a new graph in a group is viewed?

Reply:
You are right. `DiagramViewModel::rebuildDiagram()` was called too many times when it was not necessary. I updated the code - now `DiagramViewModel::rebuildDiagram()` should be only called when needed

Reply:
my changed caused the checkbox to not be updated anymore when changing the profile (thanks @chhagedorn for finding that bug). Should be fixed now.